packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -570,21 +603,37 @@ export class ConnectionManager implements IConnectionManager {
         return true;
     }
 
+    /**
+     * Cancel in-progress connection attempt.
+     */
+    private cancelConnection() {
+        assert(this.pendingConnection !== undefined, "this.pendingConnection is undefined when trying to cancel");
+        this.pendingConnection.abortController.abort();
+        this.pendingConnection = undefined;
+        this.logger.sendTelemetryEvent({ eventName: "ConnectionAttemptCancelled" });
+    }
+
     /**
      * Once we've successfully gotten a connection, we need to set up state, attach event listeners, and process
      * initial messages.
      * @param connection - The newly established connection
      */
-     private setupNewSuccessfulConnection(connection: IDocumentDeltaConnection, requestedMode: ConnectionMode) {
+     private setupNewSuccessfulConnection(
+         connection: IDocumentDeltaConnection,
+         requestedMode: ConnectionMode,
+         abortSignal: AbortSignal,
+        ) {
         // Old connection should have been cleaned up before establishing a new one
         assert(this.connection === undefined, 0x0e6 /* "old connection exists on new connection setup" */);
         assert(!connection.disposed, 0x28a /* "can't be disposed - Callers need to ensure that!" */);
 
-        if (this.pendingConnection) {
-            this.pendingConnection = false;
-        } else {
-            assert(this.closed, 0x27f /* "reentrancy may result in incorrect behavior" */);
+        this.pendingConnection = undefined;
+
+        if (this.closed || abortSignal.aborted === true) {
+            connection.dispose();
```

Comment:
There's lots of other logic in `disconnectFromDeltaStream` besides `connection.dispose`... I'm not sure about replacing the below `if (this.closed)` check with this block.  Also this skips the `this.set_readonlyPermissions` call below.

Reply:
Is there a change coming here?

Reply:
Yes, I seperated the checks for `if (abortSignal.aborted === true)` and `if (this.closed)`. In the case of the ConnectionManager being closed, it makes sense to call `this.set_readonlyPermissions()`, but in the case of a cancelled connection I think we should just end the flow here and avoid making that call.

Reply:
Gotcha. Can we just move this check to connectCore, and not pass `abortSignal` at all?  Basically, check for abort at the top of the loop and right after the loop and that's it - no one else has to worry about it.

Also, it seems like a shame to throw out the connection if the mode matches.  But probably best to keep it simple, especially given how squirrely the connection mode seems to be, ha. Probably won't happen much (I would log in this case same as the abort on top of the loop)